toast notifications(+)

mshafer1 opened the pull request:
# Why

I'm looking for a Windows client that can run as a service (or minimized to notifactions tray) -> not just minimized to the task bar so that it runs "in the background" AND that has an option to make the notifications stay up until dismissed (because I'm pushing alerts that are important and need to be seen).

ntfysh-windows already has the first, but by using the default Windows notification tooling, the ability to set the notification to stay for requested time is hampered.

This PR replaces calling the built-in Windows Toast notification for using an animated window. This provides the flexibility to have a timeout progress bar or have an infinite/until-dismissed notification.

# What I have done
Make an option to use "custom" notifications (if False - the default - still uses Windows toad notification, if true, use new notification).

That custom WinForm still:
- shows title, icon, and message
- now offers a scroll bar if the message is very long

and now:

- uses a progress bar to show a timeout for how long the message will display (can be disabled in settings)
- cancels that timeout if the dialogue is clicked on
- allows copy/pasting the title and message
- allows dismissing by click the "x" button
- if timeout is set to 0 in settings, stays open until dismissed or another message is sent.

I also introduced ntfysh_client.Themes namespace which contains a `default` (based on the current Settings dialogue and a `darkMode` to allow the user to choose the style/theming of their pop ups.

# Testing / Demo
Demo: https://youtu.be/BF3kzUK8DR0

mshafer1 committed:
make tray pop up and make it work

mshafer1 committed:
work on appearance of notification

mshafer1 committed:
get timeout working

mshafer1 committed:
implement icon

mshafer1 committed:
create a constants class to manage flag names

mshafer1 committed:
rearrange file to properties, public, protected, private

mshafer1 committed:
remove unused imports

mshafer1 committed:
if already visible, hide then show

ideally, we would stack multiple up each with its own timeout and close event. For now, this just makes a pop up for the new message.

mshafer1 committed:
use a rich text box to get scroll bar if needed and make window a little smaller

mshafer1 committed:
allow a timeout of -1 in settings

mshafer1 committed:
add a progress bar and timeout text

mshafer1 committed:
don't show notification window in task bar

alexhorner commented:
Hey there @mshafer1,

I haven't been able to do a full code review yet, but will do so tomorrow when I get the chance.

I think my main response right now would be your use case doesn't fit everyone's requirements. I for one wouldn't like the new notifications and would find them extremely frustrating and distracting, but I can absolutely see the appeal in the more obstructive and attention grabbing style.

I think a full replacement is a little too far, but a good compromise would be to add a configuration option which allows the user to choose which type of notification they would like to receive, with the native Windows one being the default as this is what users so far have come to expect from the application.

mshafer1 commented:
I think a config option is a good idea. I contemplated adding that in round
one but thought but wasn't sure if it should be included in the MVP.

I could also see a strong case for making them sized more like before, but
beyond just shrinking it, I would argue non MVP.

I also don't actually want the status/progress bar, as my use-case is
important only (i.e. would be set to not timeout).

Would you rather I take that part back out, or make that a config option as
well.

On Fri, Dec 20, 2024, 18:04 Alexander Horner ***@***.***>
wrote:

> Hey there @mshafer1 <https://github.com/mshafer1>,
>
> I haven't been able to do a full code review yet, but will do so tomorrow
> when I get the chance.
>
> I think my main response right now would be your use case doesn't fit
> everyone's requirements. I for one wouldn't like the new notifications and
> would find them extremely frustrating and distracting, but I can absolutely
> see the appeal in the more obstructive and attention grabbing style.
>
> I think a full replacement is a little too far, but a good compromise
> would be to add a configuration option which allows the user to choose
> which type of notification they would like to receive, with the native
> Windows one being the default as this is what users so far have come to
> expect from the application.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/lucas-bortoli/ntfysh-windows/pull/17#issuecomment-2557898786>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AATSJ4M2YEENZOZPDQPYPFT2GSWCPAVCNFSM6AAAAABUAAD4F2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNJXHA4TQNZYGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

alexhorner commented:
An feature MVP should strive to at least not break existing expectations or functionality for sure, so I think it's an important addition for a config option to choose.

You might wish to add an option to show or hide the counter and/or progress bar, but I think it's a good attention catching feature once again and has value.

The size and style could do with being made more consistent with the rest of the application, but for now I would say it works fine

mshafer1 committed:
add to options

mshafer1 committed:
if settings says to use the native notifications, use them

mshafer1 committed:
pass new options down to custom notifications form

mshafer1 committed:
handle if option to show progress bar is off

mshafer1 commented:
@alexhorner, all right, I got a draft of the settings dialogue updated and have tested that everything but the dark-mode switch works (that's not implemented yet). I'll plan to work on making the default be winForms default colors in the morning.

mshafer1 commented on ntfysh_client/MainForm.cs:
```diff
@@ -361,7 +397,7 @@ private void LoadSettings()
             Program.Settings = settings;
             
             //Check the settings revision. If it is older than the current latest revision, apply the settings defaults missing from previous revision
-            if (Program.Settings.Revision < defaultSettings.ReconnectAttempts)
+            if (Program.Settings.Revision < defaultSettings.Revision)
```
I noticed this while debugging.... while I didn't do a proper [Chesterton's Fence investigation](https://en.wikipedia.org/wiki/G._K._Chesterton#Chesterton's_fence) on it, it seem pretty wrong...., I'll dig into history on this more later.

alexhorner replied:
Thaaaat looks like a keyboard ninja tab tab enter move typo 🤣 Was that me? Whoops!

mshafer1 replied:
I'm going to mark this as resolved.

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+
+        private void handleTimeout(object? sender, EventArgs? e)
+        {
+            this.cancelTimer();
```
Nitpick: newline

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+
+        private Image? ConvertToolTipIconToImage(ToolTipIcon icon)
+        {
+            switch (icon)
```
This is a great candidate for the new `switch expression` syntax, and then it can be turned into a lambda

```csharp
private Image? ConvertToolTipIconToImage(ToolTipIcon icon) => icon switch
{
    ToolTipIcon.Info => SystemIcons.Information.ToBitmap(),
    ToolTipIcon.Warning => SystemIcons.Warning.ToBitmap(),
    ToolTipIcon.Error => SystemIcons.Error.ToBitmap(),
    _ => null
}
```

mshafer1 replied:
updated

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+            public const int AW_BLEND = 0x00080000;
+        }
+
+        private void window_MouseDown(object sender, EventArgs e)
```
Nitpick: Lambda one liner

mshafer1 replied:
swapped

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+                this.displayTimeoutTimer.Stop();
+                this.displayTimeoutTimer.Dispose();
+                this.displayTimeoutTimer = null;
+            }
```
Nitpick: newline

mshafer1 replied:
I added a blank line before the if statement

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+                this.updateTimer.Stop();
+                this.updateTimer.Dispose();
+                this.updateTimer = null;
+            }
```
Nitpick: newline

mshafer1 replied:
I added a blank line before the if statement

alexhorner commented on ntfysh_client/SettingsDialog.cs:
```diff
@@ -37,5 +73,20 @@ private void cancelButton_Click(object sender, EventArgs e)
             DialogResult = DialogResult.Cancel;
         }
+
+        private void setNotificationsUIElements()
```
Nitpick: Naming convention is capitalised first letter

mshafer1 replied:
renamed

alexhorner commented on ntfysh_client/SettingsDialog.cs:
```diff
@@ -37,5 +73,20 @@ private void cancelButton_Click(object sender, EventArgs e)
+            timeout.Minimum = useCustomTrayNotifications.Checked ? -1 : 0;
+        }
+
+        private void useCustomTrayNotifications_CheckedChanged(object sender, EventArgs e)
```
Nitpick: Naming convention is capitalised first letter

mshafer1 replied:
Changed

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+            this.IsVisible = false;
+        }
+
+        private void handleTimeout(object? sender, EventArgs? e)
```
Nitpick: Naming convention is capitalised first letter

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+            this.Left = workingleft - NotificationDialog.ScreenMargin;
+        }
+
+        private void ui_hide_window(object? sender, EventArgs? e)
```
Nitpick: Naming convention is capitalised first letter title casing without underscores

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+        private const int ScreenMargin = 20;
+
+        private int _timeout = 0;
+        private System.Timers.Timer? displayTimeoutTimer = null;
```
Nitpick: convention is use of _ for private variables to avoid the use of this. Goes for all others in this section

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+        private System.Windows.Forms.Timer? updateTimer = null;
+        private Stopwatch? shownStopwatch = null;
+        private ToolTipIcon? _icon;
+        private int _progress_value = 0;
```
This appears to be set but never read outside of the `progress` getter

mshafer1 replied:
removed

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+        private int _progress_value = 0;
+        private int progress
+        {
+            get { return this._progress_value; }
```
Could this just read `this.progressBar1.Value` instead?

mshafer1 replied:
yes, I originally thought there would be more to the setter. Removed.

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+        }
+
+        public bool IsVisible
+        {
```
Any reason we're wrapping this? Is it protected usually?

mshafer1 replied:
Not in the end. Removed. (I had originally though the UI would be `.SetInfo(...); .Show()` or something, but the way it is now has a better expandability and doesn't use this.

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+
+        public NotificationDialog()
+        {
+            this.ShowInTaskbar = false;
```
Does this need `this`es?

mshafer1 replied:
no, that's just what style I was used to back when I last worked in C# (removed all)

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+            {
+                // close the current notification
+                this.handleTimeout(null, null);
+            }
```
Nitpick: newline

mshafer1 replied:
I added a blank line before the if statement

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+                this.handleTimeout(null, null);
+            }
+            // setup data
+            this._icon = icon;
```
Nitpick: newline

mshafer1 replied:
I added a blank line before the if statement

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+            if (this._icon != null)
+            {
+                this.iconBox.Image = ConvertToolTipIconToImage(_icon.Value);
+            }
```
Nitpick: newline

mshafer1 replied:
I added a blank line before the if statement

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+            {
+                this.displayTimeoutTimer.Stop();
+                this.displayTimeoutTimer.Dispose();
+            }
```
Nitpick: newline

mshafer1 replied:
I added a blank line before the if statement

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+            {
+                this.updateTimer.Stop();
+                this.updateTimer.Dispose();
+            }
```
Nitpick: newline

mshafer1 replied:
I added a blank line before the if statement

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+                    this.progressBar1.Visible = true;
+                    this.lbTimeout.Visible = true;
+                    this._timeout = timeout_ms;
+                } else
```
Nitpick: newline

mshafer1 replied:
heh, wonder where that style came from...

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+            this.cancelTimer();
+        }
+
+        private void cancelTimer()
```
Nitpick: Naming convention is capitalised first letter

mshafer1 replied:
renamed

alexhorner commented on ntfysh_client/MainForm.cs:
```diff
@@ -69,8 +72,22 @@ private void OnNotificationReceive(object sender, NotificationReceiveEventArgs e
+            }
+            else
+            {
+                this.notificationDialog.ShowNotification(
```
Nitpick: newline

```csharp
this.notificationDialog.ShowNotification
(
    ...
);
```

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+            InitializeWindowHidden();
+        }
+
+        public void ShowNotification(string title, string message, int timeout_ms = -1, ToolTipIcon? icon = null, bool showTimeOutBar = true, bool showInDarkMode = true)
```
Nitpick: `timeout_ms` should be `timeoutMilliseconds`

mshafer1 replied:
This got changed to `timeoutSeconds` to deal with corresponding comment in `MainForm.cs`

alexhorner commented on ntfysh_client/MainForm.cs:
```diff
@@ -69,8 +72,22 @@ private void OnNotificationReceive(object sender, NotificationReceiveEventArgs e
+                this.notificationDialog.ShowNotification(
+                    title: finalTitle, 
+                    message: e.Message, 
+                    timeout_ms: (int)TimeSpan.FromSeconds((double)Program.Settings.Timeout).TotalMilliseconds, 
```
This seems a little cursed. If that's the best way to do it, fair enough, but if there's a better way (such as changing the type of `timeout_ms`) that might be better because we're throwing away precision.

mshafer1 replied:
I'll change it to seconds and drop the whole conversion. Only reason I made it take mS originally was to create a minimal diff with the old method.

mshafer1 replied:
interesting....

Settings.Timeout is a `decimal`, but the UI element for changing it is set to `increment=1`. Which means that the only way to get it to not be an int would be to edit the json file by hand...

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+
+        private void btnClose_Click(object sender, EventArgs e)
+        {
+            // immediate
```
immediate?

mshafer1 replied:
I'll change the comment to better explain the desired intent
```suggestion
            // don't animate, immediately "close"
```

rationale:
if a user click the x button, they expect the window to go away "now". My observation is that general Windows users are about 50% split on whether a going away animation in response to telling a window to close is annoying or "fine" (and only a few people "prefer it"). Because of this, I intentionally implemented the button close handler to immediately close the pop up (don't animate, and don't wait for the timeout).

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+            this.Opacity = 1;
+        }
+
+        private void btnClose_Click(object sender, EventArgs e)
```
I try to rename my elements to match the TitleCase naming convention. Not a major one, just a nice-to-have. I appreciate this isn't consistent within the application anyway.

mshafer1 replied:
settings dialog is using camelCase for everything....

I think the Pep8 guidance that "consistency within a file is most important" is a good point here.

I opted to make the new things in SettingsDialogue always use camelCase, but changed the NotificationDialogue be TitleCase.

alexhorner commented on ntfysh_client/SettingsDialog.cs:
```diff
@@ -37,5 +73,20 @@ private void cancelButton_Click(object sender, EventArgs e)
+        }
+
+        private const string _windowsNotificationsTimeout = "Notification Toast Timeout (seconds, may be ignored by OS based on accessibility settings):";
+        private const string _customNotificationsTimeout = "Notification Toast Timeout (seconds, use -1 to require closing notification):";
```
Would it be better to make `0` no timeout requiring closing notifications instead of `-1`? My logic behind this is `0` is not a valid value anyway, and would result in the immediate closure of the notification, which seems somewhat pointless and a bug. Then we don't end up with an invalid value between two valid values (`-1` and `1`) and can enforce a minimum value of `0` easily.

mshafer1 replied:
0 for not timeout makes sense. Although, a 0 was valid before.

(but...., If I understand correctly, a default Windows install will ignore anything outside of 1-10 and just use 10 instead, so a 0 before became 10 s, ...., but the way we have it setup now anyone upgrading will have to open the settings dialogue to switch over anyays...)

I'm leaning toward using 0.

mshafer1 replied:
Changed to 0 and dropped -1 handling/updating

alexhorner reviewed (changes requested):
HI @mshafer1

Please see attached review comments for my full code review.

Thanks for your work!

mshafer1 commented on ntfysh_client/MainForm.cs:
```diff
@@ -298,10 +323,14 @@ private void LoadTopics()
             ReconnectAttempts = 10,
-            ReconnectAttemptDelay = 3
+            ReconnectAttemptDelay = 3,
+            UseNativeWindowsNotifications = true,
```
They can. Currently these two are mutually exclusive. Preference between "UseOldNotificationsStyle" (or "New") and making it an enum? I can't imagine a 3rd notification style getting added, but FOSS is not exactly predictable. (i.e., I would lean toward the single boolean)

alexhorner replied:
Could do an enum?

mshafer1 replied:
Changed to an enum.

mshafer1 commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+            if (this.shownStopwatch == null)
+            {
+                return;
+            }
```
@alexhorner, is there a linter tool or editor config file that I can setup to check code against the desired style?

mshafer1 replied:
I went through and attempted to add a blank line before every if statement that was not the opening statement of the parent scope (and if preceded by a comment, the blank line went before the comment).

alexhorner replied:
> @alexhorner, is there a linter tool or editor config file that I can setup to check code against the desired style?

So typically I use Rider (alternatively Resharper) whilst working and it picks out most of these things. As far as I know, what it works on is the microsoft standard convention, so I just use that as it's pretty universal and sensible.

mshafer1 replied:
I installed an ran the Rider ide and looked through for additional style changes (and applied the ones that didn't seem to conflict with the previous comments).

mshafer1 committed:
use pep8 style private naming and remove this.

mshafer1 committed:
add blank lines for readability

mshafer1 commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+        }
+
+        protected override void OnShown(EventArgs e)
+        {
```
I think this is just left over from my first attempts at getting the animation to work.

mshafer1 replied:
removed

mshafer1 committed:
use CamelCasing

mshafer1 committed:
improve comment on close handler

mshafer1 committed:
more CamelCasing

mshafer1 committed:
utilize modern c# features

mshafer1 committed:
just utilize the progress bar directly

mshafer1 committed:
remove IsVisible and just use Visible instead

mshafer1 committed:
remove _icon

mshafer1 committed:
format

mshafer1 committed:
rename new elements to be camelCase to match existing

mshafer1 committed:
rename elements in new dialogue to PerlCase per PR comment

mshafer1 committed:
fix button hover behavior

mshafer1 committed:
introduce theme files and handle in notification popup

mshafer1 committed:
remove unused imports

mshafer1 committed:
use 0 for "don't timeout" and don't allow -1

mshafer1 committed:
use seconds for value of timeout rather then mS

mshafer1 commented on ntfysh_client/MainForm.cs:
```diff
@@ -16,6 +16,7 @@ public partial class MainForm : Form
         private readonly NotificationListener _notificationListener;
         private bool _startInTray;
         private bool _trueExit;
+        private NotificationDialog notificationDialog;
```
I tried to rename all new private fields to `_...` and remove all occurrences of `this.`. It seems to be working.

mshafer1 commented on ntfysh_client/MainForm.cs:
```diff
@@ -312,6 +341,13 @@ private void MergeSettingsRevisions(SettingsModel older, SettingsModel newer)
                 older.ReconnectAttempts = newer.ReconnectAttempts;
                 older.ReconnectAttemptDelay = newer.ReconnectAttemptDelay;
             }
+            if (older.Revision < 2)
```
updated

mshafer1 commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+
+        private void UpdateProgress(object? sender, EventArgs e)
+        {
+            if (this.shownStopwatch == null)
```
changed

mshafer1 commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+
+        protected override void SetVisibleCore(bool value)
+        {
+            this.SetWindowPosition();
```
I added a blank line before the if statement

mshafer1 commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+                    time: 250,
+                    flags: NFWinUserAnimateWindowConstnats.AW_SLIDE | NFWinUserAnimateWindowConstnats.AW_VER_NEGATIVE
+                );
+            }
```
I added a blank line before the if statement

mshafer1 commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+            if (value)
+            {
+                this.BringToFront();
+                AnimateWindow(
```
Added

mshafer1 commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+
+        private void ui_hide_window(object? sender, EventArgs? e)
+        {
+            AnimateWindow(
```
added

mshafer1 commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -0,0 +1,263 @@
+                this.Handle,
+                time: 250,
+                flags: NFWinUserAnimateWindowConstnats.AW_SLIDE | NFWinUserAnimateWindowConstnats.AW_VER_POSITIVE | NFWinUserAnimateWindowConstnats.AW_HIDE
+            );
```
I added a blank line before the call to the parent class

mshafer1 reviewed (commented):
I think I've addressed all current comments.

I also added in "themes" and implemented the "showCustomNotificationsInDarkMode" switch.

There are some open questions still. I'll update the demo/screen-recording once those are addressed.

mshafer1 committed:
change default of new notifications to light mode

mshafer1 committed:
massage the colors a little more

mshafer1 committed:
use an enum

mshafer1 committed:
apply Rider suggestions

mshafer1 committed:
Make the blue a little closer to the true default

mshafer1 commented:
@alexhorner, are you accustomed to "reviewer marks conversations as resolved once they confirm they have been addressed" or to "submitter marks conversations as resolved once they have attempted to address them"?

alexhorner commented:
I wouldn't say either to be honest. As long as they remain visible and I can go back and see them I guess you can mark them as resolved once you're satisfied with the solution you have produced.

mshafer1 commented:
> I wouldn't say either to be honest. As long as they remain visible and I can go back and see them I guess you can mark them as resolved once you're satisfied with the solution you have produced.

GitHub kind of tries to hide resolved conversations....

alexhorner commented:
> > I wouldn't say either to be honest. As long as they remain visible and I can go back and see them I guess you can mark them as resolved once you're satisfied with the solution you have produced.
> 
> GitHub kind of tries to hide resolved conversations....

Ah that's a pain. I'm not used to using GitHub code review tools. Too used to DevOps

mshafer1 commented:
@alexhorner, do you need anything else from me before doing another round of review?

alexhorner commented:
@mshafer1 Looks fine to me at this point. One bug I have noticed in testing is that in native windows notification mode, the ping sound still plays. This is unnecessary as Windows plays its own sound.

mshafer1 commented:
Heh, I hadn't thought of sound.

Let me check on that.

On Thu, Jan 9, 2025, 15:59 Alexander Horner ***@***.***>
wrote:

> @mshafer1 <https://github.com/mshafer1> Looks fine to me at this point.
> One bug I have noticed in testing is that in native windows notification
> mode, the ping sound still plays. This is unnecessary as Windows plays its
> own sound.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/lucas-bortoli/ntfysh-windows/pull/17#issuecomment-2581322226>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AATSJ4P25TTEXIUWVZ6BODT2J3WMDAVCNFSM6AAAAABUAAD4F2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOBRGMZDEMRSGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mshafer1 committed:
add support for still playing the Windows alert sound

mshafer1 commented:
> Heh, I hadn't thought of sound. Let me check on that.
> […](#)
> On Thu, Jan 9, 2025, 15:59 Alexander Horner ***@***.***> wrote: @mshafer1 <https://github.com/mshafer1> Looks fine to me at this point. One bug I have noticed in testing is that in native windows notification mode, the ping sound still plays. This is unnecessary as Windows plays its own sound.

@alexhorner, I have added a setting for "custom notifications play default Windows sound" and implemented and tested it.

alexhorner commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -100,6 +102,10 @@ public void ShowNotification(string title, string message, int timeoutSeconds =
             // ok, show the window
             Show();
             SetWindowPosition();
+            if(playNotificationSound)
```
Nitpick: One liner & newline

```csharp
SetWindowPosition();

if (playNotificationSound) PlayNotificationSound();
```

mshafer1 replied:
applied

alexhorner reviewed (changes requested):
@mshafer1 I'm a fool! The ping sound was coming from the ntfy.sh website I was using during testing. My comment must've been a bit confusing lol, I thought you implemented a custom sound for your notifications too. which played at the same time as the Windows sound, but I see now that you actually just didn't play any sound.

Either way, your latest commit looks good and fixes the real issue, not the one I mistook it for 😛 I have tested and provided a little more review feedback but other than that I think this is a nice addition ready to make its way in

mshafer1 committed:
apply suggestions

mshafer1 committed:
focus the close button before first launch and on every show to make title text consistently not selected

mshafer1 commented on ntfysh_client/NotificationDialog.cs:
```diff
@@ -265,5 +271,28 @@ private void CancelTimer()
             }
         }
+
+        private void PlayNotificationSound()
```
Thanks

alexhorner commented:
> @alexhorner I was quite surprised when I unmuted my laptop and heard a bell noise on the custom :) (things made a little more sense to me when I switched back to default and also heard the bell -> I wound up muting the site's tab while testing :D)
> 
> I have addressed your feedback, and pushed one more commit to deal with the title text continuing to be selected on my machine.

Aha! We both got caught by that one did we? 😆 

All looks good to me, thank you for your hard work!

@lucas-bortoli This one should be ready to merge, I have tested. As an aside, would you consider adding me as a maintainer so I may accept pull requests and package and release binaries on the releases page? Thanks!

mshafer1 commented:
@lucas-bortoli, any updates? I would like to deploy this soon and would prefer to do some with a build from the shared main branch.

alexhorner commented:
@lucas-bortoli Bump
